scripts/lmdbg-stat.1.in
.\" Copyright (c) 2010-2011 by Aleksey Cheusov (vle@gmx.net)
.\" Absolutely no warranty.
.\"
.\" ------------------------------------------------------------------
.de VB \" Verbatim Begin
.ft CW
.nf
.ne \\$1
..
.de VE \" Verbatim End
.ft R
.fi
..
.\" ------------------------------------------------------------------
.TH LMDBG-STAT 1 "Nov 26, 2011" "" ""
.SH NAME
lmdbg-stat \- outputs statistical information based on lmdbg-run output
.SH SYNOPSIS
.BI lmdbg-stat " [OPTIONS] [files...]"
.SH DESCRIPTION
.B lmdbg-stat
takes output of
.B lmdbg-run
or other
.B lmdbg-*
utilities on input
and outputs total and per-stacktrace statistical information about
memory allocations.
The following statistics is output:
.RS
.TP
.IR allocs " or " total_allocs
Per-stacktrace or total number of memory allocations
.TP
.IR max " or " total_max
Per-stacktrace or total maximum of requested bytes
.TP
.IR peak " or " total_peak
Per-stacktrace or total peak of allocated memory
.TP
.IR leaks " or " total_leaks
Per-stacktrace or total amount of bytes that were allocated but not freed
.RE
.SH OPTIONS
.TP
.B "-h"
Display the help message.
.TP
.BI "-V"
Display the lmdbg version.
.SH EXAMPLE
.VB
$ cat -n testme.c
     1  #include <stdlib.h>
     2  
     3  int main ()
     4  {
     5     int i;
     6     void *p1, *p2;
     7     p1 = malloc (100);
     8     for (i=0; i < 3; ++i){
     9        p2 = realloc (NULL, 10+30*i);
    10     }
    11     free (p1);
    12     free (p2);
    13  
    14     return 0;
    15  }
$ cc -O0 -g -o testme testme.c
$ lmdbg-run -o log -f 'lmdbg-stat | lmdbg-sym' ./testme     
$ cat log
info stat total_leaks: 50
info stat total_allocs: 4
info stat total_free_cnt: 2
stacktrace peak: 120 max: 70 allocs: 3 leaks: 50
 0xbbbe2bc3     lmdbg.c:101     log_stacktrace
 0xbbbe3498     lmdbg.c:456     realloc
 0x8048788      testme.c:9      main
 0x80485b4      
 0x8048517      
stacktrace peak: 100 max: 100 allocs: 1
 0xbbbe2bc3     lmdbg.c:101     log_stacktrace
 0xbbbe33bd     lmdbg.c:431     malloc
 0x8048757      testme.c:7      main
 0x80485b4      
 0x8048517      
$ 
.VE
.SH SEE ALSO
.BR lmdbg(1) ,
.BR lmdbg-run(1) ,
.BR lmdbg-sym(1) ,
.BR lmdbg-sort(1) ,
.BR lmdbg-grep(1) ,
.BR lmdbg-leaks(1) ,
.BR lmdbg-sysleaks(1) ,
.BR lmdbg-strip(1) ,
.B lmdbg-modules(1)
.SH AUTHOR
Aleksey Cheusov <vle@gmx.net>


NEWS
======================================================================
Version 0.17.0, by Aleksey Cheusov, Thu, 24 Nov 2011 00:09:48 +0300

  lmdbg-sym:
    - Doesn't accept progname as an argument anymore.
      Use -P option instead.
    - option --with-so was replaced with -s.

  Long options have been removed from all programs.
  use standard single-letter POSIX-compatible options.

  SORRY FOR INCOMPATIBILITIES WITH PREVIOUS RELEASES.

                                    ----------------------------------

  lmdbg-stat, tests/test.sh:
    - Fixes for Solaris

  lmdbg-run:
    - options -T and -B were added to lmdbg-run.
    - option -n was added. It disables logging on startup.
      Sending SIGUSR1 signal to the process enables it.

  lmdbg:
    - -M option was added
    - -n option was added
    - -T and -B options were added
    - Regression tests were added for -M

  lmdbg-grep.1:
    - documentation for "bytes" variable

  If set, LMDBG_MODULES_AWK and LMDBG_STRIP_AWK environment variables
  are used as a path to AWK interpreter by "lmdbg-modules" and
  "lmdbg-strip" programs.

  More regression tests

======================================================================
Version 0.16.0, by Aleksey Cheusov, Thu, 16 Dec 2010 17:54:22 +0200

  Spelling corrections in man pages

  lmdbg-stat:
    - reimplemented in C. Now it uses libJudy library.
      See INSTALL for details.

  lmdbg-run:
    - NULL is *always* output as string "NULL"

======================================================================
Version 0.15.1, by Aleksey Cheusov, Thu,  7 Oct 2010 22:22:16 +0300

  FIX: On Linux/glibc systems lmdbg-run may fail (assert(3))
  because calloc(3) actually uses malloc(3). So, logging calloc(3)
  has been disabled.

  mk-configure >= 0.20 is required for building.

  Minor fixes and clean-ups.

======================================================================
Version 0.15.0, by Aleksey Cheusov, Wed, 26 May 2010 16:17:27 +0300

  Latest "mk-configure" is required for building.

  External tool "runawk" is needed for newly created "lmdbg-strip" and
  "lmdbg-modules" programs. http://sourceforge.net/projects/runawk/

  New utility "lmdbg-module" for enriching stacktraces given on input
  with information about module by set of rules specified by user.

  New utility "lmdbg-strip" for sripping stacktrace given on input.

  lmdbg-sort:
     - new sorting field "num"
     - minor fix in sorting order (Linux + gawk)
     - support for "info modulestat" input

  lmdbg-grep:
     - new variables to check: "op", "op_type" and "bytes".
     - new field to grep: "module".

  lmdbg-stat:
     - accepts "num: <num>" per-stacktrace information on input

  lmdbg:
     -ofilename is accepted and is equal to '-o filename'.

  lmdbg-run:
     - -ppipe is accepted and is equal to '-p pipe'.
     - by default 'num: <number>' tokens are output for each stacktrace.

  lmdbg-m2s, lmdbg-s2m:
    Internal singleline format has changed. As a result lmdbg-m2s|lmdbg-s2m
    pair works correctly with input that doesn't have real stacktraces.

  Fix in usage message for most lmdbg-* programs.

  Fixes for warnings messages produced by gcc.

  More regression tests.

======================================================================
Version 0.14.0, by Aleksey Cheusov, Tue, 13 Apr 2010 17:06:26 +0300

  New tools: lmdbg-stat, lmdbg-grep and lmdbg-sort for collecting and
  analysing statistical information about memory allocations.

  lmdbg-run: new options for lmdbg-run: -S and -M for generating
  shortened stacktraces.

  lmdbg-sym: new option -p for obtaining program name from lmdbg-run's
  output.

  fix: 'mkcmake test' remove its temporary files.

======================================================================
Version 0.13.0, by Aleksey Cheusov, Sun, 31 Jan 2010 18:40:05 +0200

  INSTALL: updated

  mk-configure>=0.14.0 is needed to build lmdbg.

  FIXED: 'make install' command installs test utilities from tests/
  subdirectory. Thanks to Mikolaj Golub for bug report.

  lmdbg-sym:
    - LMDBG uses very tricky features of GDB. As a result GDB may
      segfault on some platforms, e.g. gdb-6.4.90-debian fails on
      Debian/Etch.  In order to show users that GDB is buggy and needs
      to be fixed, an exit status of it is checked. For this purpose
      http://sf.net/projects/pipestatus/ is needed.
    - LMDBG_GDB environment variable sets the path to gdb program
      overiding path defined at build time. See scripts/Makefile.

  For better efficiency lmdbg-multi2single and lmdbg-single2multi
  internal scripts have been reimplemented in C and renamed to
  lmdbg-m2s and lmdbg-s2m respectively.

  Built-time AWK variable sets an AWK interpreter for awk script.
  See scripts/Makefile.

======================================================================
Version 0.12.0, by Aleksey Cheusov, Sun,  3 Jan 2010 14:05:15 +0200

  ADDED: support for calloc(3) and posix_memalign(3).

  Latest mk-configure is required for building.

  A target for testing lmdbg is again "test"! :-)

  Partial support for OpenBSD. Partial because I don't know how to
  detect where shared lib's code begins, there are no /proc/$$/maps.

  added: -a and -g options for "lmdbg" script.

  cleanups:
    - SHLIB_MAJOR/MINOR are used instead of dirty hacks.

  "make test" is now friendly with MAKEOBJDIR

  More regression tests.

======================================================================
Version 0.11.0, by Aleksey Cheusov, Sat, 25 Jul 2009 01:48:15 +0300

  Support for modern Linux/glibc where libraries are loaded
  to different regions of the virtual memory
  (See http://en.wikipedia.org/wiki/Address_space_layout_randomization).
  To handle this lmdbg-run adds the special "info section" information
  obtained from /proc/<pid>/maps available on NetBSD and Linux
  to the log file, and lmdbg-sym uses this information.

  fixed: enabling malloc logging in lmdbg-sym was an extreamly strange

stat/lmdbg-stat.c
/*
  Copyright (c) 2003-2009 Aleksey Cheusov <vle@gmx.net>

  Permission is hereby granted, free of charge, to any person obtaining
  a copy of this software and associated documentation files (the
  "Software"), to deal in the Software without restriction, including
  without limitation the rights to use, copy, modify, merge, publish,
  distribute, sublicense, and/or sell copies of the Software, and to
  permit persons to whom the Software is furnished to do so, subject to
  the following conditions:

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>

#include <Judy.h>

#include "st_hash.h"
#include "stat.h"

static int line_num = 0;

static void *hash = NULL;

static void xputc (int c, FILE *stream)
{
	if (putc (c, stdout) == EOF){
		perror ("putc(3) failed");
		exit (1);
	}
}

static void **stacktrace    = NULL;
static int  stacktrace_len = 0;
static int  stacktrace_max = 0;

typedef enum {
	ft_malloc=10,
	ft_realloc,
	ft_memalign,
	ft_posix_memalign,
	ft_free
} func_t;

typedef struct {
	func_t type;
	size_t bytes;
	void *oldaddr;
	void *addr;
} operation_t;

static operation_t op;

static int total_allocs_cnt = 0;
static int total_free_cnt = 0;
static size_t total_allocated = 0;

typedef struct {
	size_t allocated;
	int stacktrace_id;
} ptrdata_t;

static Pvoid_t ptr2data = NULL;

static void ptr2data_destroy (void)
{
	Word_t idx = 0;
	stat_t ** ptr = (stat_t **) JudyLFirst (ptr2data, &idx, NULL);
	if (!ptr)
		return;

	do {
		if (*ptr)
			free (*ptr);

		ptr = (stat_t **) JudyLNext (ptr2data, &idx, NULL);
	} while (ptr != NULL);
	JudyLFreeArray (&ptr2data, NULL);
}

static void **stacktrace_dup (void **st, int st_len)
{
	void **p = malloc (st_len * sizeof (st [0]));
	memcpy (p, st, st_len * sizeof (st [0]));
	return p;
}

static void process_stacktrace (void)
{
	PWord_t ret;
	int id;
	int old_maxid;
	int new = 0;
	ptrdata_t *ptrdata;
	stat_t **stat_cell_p = NULL;
	stat_t *stat_cell = NULL;

	if (!stacktrace_len)
		return;

	switch (op.type){
		case ft_malloc:
		case ft_realloc:
		case ft_memalign:
		case ft_posix_memalign:
			old_maxid = st_hash_getmaxid (hash);
			id = st_hash_insert (hash, stacktrace, stacktrace_len);

			new = id > old_maxid;
//			if (new){
//				statistics = (stat_t*) realloc (
//					statistics,
//					sizeof (statistics [0]) * (id + 1));
//				memset (&statistics [id],0, sizeof (statistics [0]));
//			}

			stat_cell_p = get_stat (id);
			if (!*stat_cell_p){
				*stat_cell_p = malloc (sizeof (stat_t));
				memset (*stat_cell_p, 0, sizeof (stat_t));
			}
			stat_cell = *stat_cell_p;
			stat_cell->allocated += op.bytes;
			if (stat_cell->allocated > stat_cell->peak_allocated)
				stat_cell->peak_allocated = stat_cell->allocated;
			if (op.bytes > stat_cell->max_allocated)
				stat_cell->max_allocated = op.bytes;
			++stat_cell->allocs_cnt;
			if (new){
				stat_cell->stacktrace
					= stacktrace_dup (stacktrace, stacktrace_len);
				stat_cell->stacktrace_len = stacktrace_len;
			}
			++total_allocs_cnt;
			total_allocated += op.bytes;

			/* pointer to associated data */
			ret = (PWord_t) JudyLIns (&ptr2data, (Word_t) op.addr, 0);
			*ret = (Word_t) malloc (sizeof (ptrdata_t));
			ptrdata = (ptrdata_t*) *ret;
			ptrdata->allocated = op.bytes;
			ptrdata->stacktrace_id = id;
			break;
		case ft_free:
			break;
		default:
			abort ();
	}

	switch (op.type){
		case ft_malloc:
		case ft_memalign:
		case ft_posix_memalign:
			break;
		case ft_realloc:
			if (op.oldaddr){
				ptrdata = *(ptrdata_t **) JudyLGet (
					ptr2data, (Word_t) op.oldaddr, 0);
				total_allocated -= ptrdata->allocated;
				stat_cell_p = get_stat (ptrdata->stacktrace_id);
				stat_cell = *stat_cell_p;
				stat_cell->allocated -= ptrdata->allocated;

				free (ptrdata);
				JudyLDel (&ptr2data, (Word_t) op.oldaddr, 0);
			}
			break;
		case ft_free:
			if (op.oldaddr){
				ptrdata = *(ptrdata_t **) JudyLGet (
					ptr2data, (Word_t) op.oldaddr, 0);
				total_allocated -= ptrdata->allocated;
				stat_cell_p = get_stat (ptrdata->stacktrace_id);
				stat_cell = *stat_cell_p;
				stat_cell->allocated -= ptrdata->allocated;

				++total_free_cnt;

				free (ptrdata);
				JudyLDel (&ptr2data, (Word_t) op.oldaddr, 0);
			}
			break;
		default:
			abort ();
	}

	/* reinit */
	stacktrace_len = 0;
}

static void print_results (void)
{
	int i, j;
	int st_count;
	stat_t *stat_cell;
	char buf [200];

	printf ("info stat total_allocs: %i\n", total_allocs_cnt);
	printf ("info stat total_free_cnt: %i\n", total_free_cnt);
	printf ("info stat total_leaks: %lu\n", (unsigned long) total_allocated);

	st_count = st_hash_getmaxid (hash);
	for (i=1; i <= st_count; ++i){
		stat_cell = *get_stat (i);

		printf ("stacktrace peak: %lu max: %lu allocs: %i",
				(unsigned long) stat_cell->peak_allocated,
				(unsigned long) stat_cell->max_allocated,
				                stat_cell->allocs_cnt);
		if (stat_cell->allocated){
			printf (" leaks: %lu", (unsigned long) stat_cell->allocated);
		}
		printf ("\n");
		for (j=0; j < stat_cell->stacktrace_len; ++j){
			buf [0] = 0;
			snprintf (buf, sizeof (buf), "%p", stat_cell->stacktrace [j]);
			if (buf [0] != '0' || buf [1] != 'x'){
				/* Under Solaris %p doesn't print 0x prefix for address */
				printf (" 0x%s\n", buf);
			}else{
				printf (" %s\n", buf);
			}
		}
	}
}

static void * s2p (const char *s)
{
	void *addr = NULL;
	int ret = 0;

	if (strcmp (s, "NULL")){
		ret = sscanf (s, "%p", &addr);
		if (!addr){
			/* Under Solaris %p doesn't accept 0x prefix for address */
			assert (s[0] == '0' && s[1] == 'x');
			ret = sscanf (s+2, "%p", &addr);
		}
		if (ret != 1){
			fprintf (stderr, "Bad address: %s\n", s);
			exit (1);
		}
	}

	return addr;
}

static int s2i (const char *s)
{
	int val;
	int ret = 0;

	ret = sscanf (s, "%i", &val);
	if (ret != 1){
		fprintf (stderr, "Bad integer: %s\n", s);
		exit (1);
	}

	return val;
}

static void process_line (char *buf)
{
	void *addr;
	char *p, *last_p;
	char * tokens [10];
	int token_count = 0;

	char orig_buf [20480];

	snprintf (orig_buf, sizeof (orig_buf), "%s", buf);

	if (!buf [0])
		return;

	if (buf [0] == ' '){
		/* address with leading space character */
		addr = s2p (buf + 1);

		++stacktrace_len;
		if (stacktrace_len > stacktrace_max){
			stacktrace_max = stacktrace_len;
			stacktrace = realloc (stacktrace,
								  stacktrace_max * sizeof (void *));
		}
		stacktrace [stacktrace_len-1] = addr;
	}else{
		process_stacktrace ();

		/* malloc/calloc/realloc/free/... */
		for (p=buf; *p; ++p){
			if (*p == ' ' || *p == '\t')
				*p = 0;
		}
		last_p = p;

		for (p=buf; p <= last_p; ++p){
			if (*p && (p == buf || p [-1] == 0)){
				tokens [token_count++] = p;
				if (token_count == sizeof (tokens)/sizeof (tokens [0])){
					/* too much tokens, ignore the rest */
					break;
				}
			}
		}

		if (token_count >= 5 &&
			!strcmp (tokens [0], "malloc") &&
			!strcmp (tokens [1], "(") &&
			!strcmp (tokens [3], ")") &&
			!strcmp (tokens [4], "-->"))
		{
			op.type    = ft_malloc;
			op.bytes   = (size_t) s2i (tokens [2]);
			op.oldaddr = NULL;
			op.addr    = s2p (tokens [5]);
			return;
		}

		if (token_count >= 8 &&
			!strcmp (tokens [0], "calloc") &&
			!strcmp (tokens [1], "(") &&
			!strcmp (tokens [3], ",") &&
			!strcmp (tokens [5], ")") &&
			!strcmp (tokens [6], "-->"))
		{
			op.type    = ft_malloc;
			op.bytes   = (size_t) (s2i (tokens [2]) * s2i (tokens [4]));
			op.oldaddr = NULL;
			op.addr    = s2p (tokens [7]);
			return;
		}

		if (token_count >= 8 &&
			!strcmp (tokens [0], "realloc") &&
			!strcmp (tokens [1], "(") &&
			!strcmp (tokens [3], ",") &&
			!strcmp (tokens [5], ")") &&
			!strcmp (tokens [6], "-->"))
		{
			op.type    = ft_realloc;
			op.bytes   = (size_t) s2i (tokens [4]);
			op.oldaddr = s2p (tokens [2]);
			op.addr    = s2p (tokens [7]);
			return;
		}

		if (token_count >= 8 &&
			(!strcmp (tokens [0], "memalign") ||
			 !strcmp (tokens [0], "posix_memalign")) &&
			!strcmp (tokens [1], "(") &&
			!strcmp (tokens [3], ",") &&
			!strcmp (tokens [5], ")") &&
			!strcmp (tokens [6], "-->"))
		{
			op.type    = ft_memalign;
			op.bytes   = (size_t) s2i (tokens [4]);
			op.oldaddr = NULL;
			op.addr    = s2p (tokens [7]);
			return;
		}

		if (token_count >= 4 &&
			!strcmp (tokens [0], "free") &&
			!strcmp (tokens [1], "(") &&
			!strcmp (tokens [3], ")"))
		{
			op.type    = ft_free;
			op.bytes   = 0;
			op.oldaddr = s2p (tokens [2]);
			op.addr    = NULL;
			return;
		}

		if (token_count >= 1 && !strcmp (tokens [0], "info")){
			puts (orig_buf);
			return;
		}

		fprintf (stderr, "bad input line: %s\n", orig_buf);
	}
}

static void process_stream (FILE *in)
{
	char buffer [10000];
	size_t len;
	while (fgets (buffer, sizeof (buffer), in)){
		len = strlen (buffer);
		if (len > 0 && buffer [len-1] == '\n')
			buffer [len-1] = 0;

		++line_num;
		process_line (buffer);
	}

	if (ferror (in)){
		perror ("fgets(3) failed");
		exit (1);
	}

	process_stacktrace ();

	if (stacktrace)
		free (stacktrace);
	stacktrace = NULL;
}

extern char *optarg;
extern int optind;

static void show_version (void)
{
	printf ("lmdbg-stat " VERSION "\n");
	exit (0);
}

static void show_help (void)
{
	printf ("\
Taking an output of lmdbg-run or other lmdbg-* utilities on input\n\
lmdbg-stat outputs a total and per-stacktrace statistical information\n\
about memory alocations.\n\
\n\
usage: lmdbg-stat [OPTIONS] [files...]\n\
OPTIONS:\n\
  -h          display this screen\n\
  -V          display version\n\
");
}

int main (int argc, char **argv)
{
	int i;
	FILE *fd;
	int opt;

	while (opt = getopt (argc, argv, "hV"), opt != -1){
		switch (opt) {
			case 'h':
				show_help ();
				exit (0);
				break;
			case 'V':
				show_version ();
				exit (0);
				break;
			case '?':
			default:
				show_help ();
				exit (1);
		}
	}

	argc -= optind;
	argv += optind;

	hash = st_hash_create ();

	if (!argc){
		process_stream (stdin);
	}else{
		for (i=0; i < argc; ++i){
			fd = fopen (argv [i], "r");
			if (!fd){
				perror ("fopen(3) failed");
				exit (1);
			}

			process_stream (fd);
			fclose (fd);
		}
	}

	print_results ();

	if (line_num >= 1)
		xputc ('\n', stdout);

	st_hash_destroy (hash);
	destroy_stats ();
	ptr2data_destroy ();

	return 0;
}
